### ISSUE TITLE
symfony/messager: how to set-up Redis transport connection by the socket?

### ISSUE THREAD

**Issue by alebedev80** [symfony/symfony#34682]
I need to set-up connection to the Redis service by the socket. How to set-up DSN?

By the code allows only DSN with valid url ([Connection.php|https://github.com/symfony/messenger/blob/master/Transport/RedisExt/Connection.php]):

```
        if (false === $parsedUrl = parse_url($dsn)) {
            throw new InvalidArgumentException(sprintf('The given Redis DSN "%s" is invalid.', $dsn));
        }
```

Is there any reason for that or is it an architectural flaw?

**Comment by chalasr** [comment]
Sounds good to me. Up for a PR?

**Comment by mertingen** [comment]
I am drastically interested in figuring it out. I totally got the problem and feature. Likewise, what way should I follow?

**Comment by flip111** [comment]
Seems that "DSN" ([Data source name](https://en.wikipedia.org/wiki/Data_source_name) aka "connection string") is not a specified format with a RFC behind it. To come up with such a format is quite some effort. Or even if you take the a format that already exist there is still the problem of parsing it correctly. Maybe better to rename the `fromDsn` function to `fromUrl` and then have another function `fromSocket` that just checks if it's a valid file on your system.

**Comment by mertingen** [comment]
Yes, I could check if it's a valid file on the system but how can it connect by using socket file on rest of the code?

**Comment by flip111** [comment]
It depends on the underlying driver ...

for example this driver https://github.com/phpredis/phpredis#example-1

**Comment by chalasr** [comment]
The Cache component handles such DSNs (Lock does probably as well, I didn't check), doing the same for Messenger seems consistent.
Implementation-wise, I suggest to take inspiration from the Cache's RedisTrait. 
The code area that needs to be patched is the `Connection` class from the Redis transport, and the related framework configuration probably needs to be updated accordingly.

**Comment by flip111** [comment]
Perhaps there is an opportunity here to avoid some duplication of code in these two files

https://github.com/symfony/symfony/blob/master/src/Symfony/Component/Messenger/Transport/RedisExt/Connection.php

https://github.com/symfony/symfony/blob/master/src/Symfony/Component/Cache/Traits/RedisTrait.php

**Comment by JJarrie** [comment]
The Lock component use an already connected Redis client.

Basically to resolve it we need, use this part of RedisTrait (The scheme part are util?) to avoid error the error of parse_url:

`
if (0 === strpos($dsn, 'redis:')) {
            $scheme = 'redis';
        } elseif (0 === strpos($dsn, 'rediss:')) {
            $scheme = 'rediss';
        } else {
            throw new InvalidArgumentException(sprintf('Invalid Redis DSN: %s does not start with "redis:" or "rediss".', $dsn));
        }

        $dsn = preg_replace_callback('#^'.$scheme.':(//)?(?:(?:[^:@]*+:)?([^@]*+)@)?#', function ($m) use (&$auth) {
            if (isset($m[2])) {
                $auth = $m[2];
            }

            return 'file:'.($m[1] ?? '');
        }, $dsn);
`

And if we are this case, use the 'path' value (of parse_rul) and avoid to pass the port to method Redis::connect.

**Comment by JJarrie** [comment]
So I have push a little too fast, this issue have to be fix on master (feature)?

**Comment by alebedev80** [comment]
Thanks!
